aeon/dj_pipeline/analysis/block_analysis.py

```diff
@@ -798,3 +798,116 @@ class AnalysisNote(dj.Manual):
+    subject_patch_data = (BlockSubjectAnalysis.Patch() & key).fetch(format="frame")
+    subject_patch_data.reset_index(level=["experiment_name"], drop=True, inplace=True)
+    subject_names = subject_patch_data.index.get_level_values("subject_name").unique()
+    wheel_ts = (BlockAnalysis.Patch & key).fetch("wheel_timestamps")[0]
+    # For each subject:
+    #   - Create cumulative wheel distance spun sum df combining all patches
+    #     - Columns: timestamp, wheel distance, patch
+    #   - Discretize into 'possible foraging events' based on `max_inactive_time`, and `min_wheel_movement`
+    #   - Filter out events with < `min_pellets`
+    #   - For final events, get: duration, n_pellets, cum_wheel_distance -> add to returned DF
+    bout_data = pd.DataFrame(columns=["start", "end", "n_pellets", "cum_wheel_dist", "subject"])
+    for subject in subject_names:
+        cur_subject_data = subject_patch_data.xs(subject, level="subject_name")
+        # Create combined cumulative wheel distance spun: ensure equal length wheel vals across patches
+        wheel_vals = cur_subject_data["wheel_cumsum_distance_travelled"].values
+        min_len = min(len(arr) for arr in wheel_vals)
+        comb_cum_wheel_dist = np.sum([arr[:min_len] for arr in wheel_vals], axis=0)
+        wheel_ts, comb_cum_wheel_dist = (  # ensure equal length wheel vals and wheel ts
+            arr[: min(len(wheel_ts), len(comb_cum_wheel_dist))] for arr in [wheel_ts, comb_cum_wheel_dist]
+        )
+        # For each wheel_ts, get the correspdoning patch that was spun
+        patch_spun = np.empty(len(wheel_ts), dtype="<U20")
+        patch_spun[:] = ""
+        wheel_spun_thresh = 0.03  # threshold for wheel movement (cm)
+        for _, row in cur_subject_data.iterrows():
+            patch_name = row.name[1]
+            diff = np.diff(row["wheel_cumsum_distance_travelled"], prepend=0)
+            spun_indices = np.where(diff > wheel_spun_thresh)[0]
+            patch_spun[spun_indices] = patch_name
+        patch_spun_df = pd.DataFrame(index=wheel_ts, columns=["cum_wheel_dist", "patch_spun"])
+        patch_spun_df["cum_wheel_dist"] = comb_cum_wheel_dist
+        patch_spun_df["patch_spun"] = patch_spun
+        wheel_s_r = pd.Timedelta(wheel_ts[1] - wheel_ts[0], unit="ns")
+        win_len = int(max_inactive_time / wheel_s_r)
+        # Find times when foraging
+        max_windowed_wheel_vals = (
+            patch_spun_df["cum_wheel_dist"]
+            .shift(-(win_len - 1))
+            .rolling(window=win_len, min_periods=1)
+            .max()
+        )
+        foraging_mask = max_windowed_wheel_vals > (patch_spun_df["cum_wheel_dist"] + min_wheel_movement)
+        # Discretize into foraging bouts
+        bout_start_indxs = np.where(np.diff(foraging_mask.astype(int), prepend=0) == 1)[0]
+        bout_end_indxs = np.where(np.diff(foraging_mask.astype(int), prepend=0) == -1)[0]
```

Comment:
Need to account for final pellet delivery in bout: time for bonsai to deliver pellet / beambreak to occur after threshold crossing

Reply:
`bout_start_indxs` shift is done in a96231ce97d72f52a6a119b17eab91c3ca082421 and left a TODO note for changing the constant for `bout_end_indxs`